include/enrico/comm.h

```diff
@@ -247,20 +263,114 @@ void Comm::send_and_recv(xt::xtensor<T, N>& values, int dest, int source) const
+    std::vector<size_t> source_shape(my_shape);
+    send_and_recv(source_shape, dest, source);
+    if (rank == dest && my_shape != source_shape) {
+      values.resize(source_shape);
     }
-    // Send the size
-    auto n = values.size();
 
     // Finally, send data
     int tag = source;
     if (rank == source) {
-      MPI_Send(values.data(), n, get_mpi_type<T>(), dest, tag, comm);
+      MPI_Send(values.data(), values.size(), get_mpi_type<T>(), dest, tag, comm);
     } else if (rank == dest) {
-      MPI_Recv(values.data(), n, get_mpi_type<T>(), source, tag, comm, MPI_STATUS_IGNORE);
+      MPI_Recv(values.data(),
+               values.size(),
+               get_mpi_type<T>(),
+               source,
+               tag,
+               comm,
+               MPI_STATUS_IGNORE);
+    }
+  }
+}
+
+template<typename T>
+std::enable_if_t<std::is_scalar<std::decay_t<T>>::value>
+Comm::send_and_recv(T& recvbuf, int dest, T& sendbuf, int source) const
+{
+  if (this->active()) {
+    if (dest != source) {
+      int tag = source;
+      if (rank == source) {
+        MPI_Send(&sendbuf, 1, get_mpi_type<T>(), dest, tag, comm);
+      } else if (rank == dest) {
+        MPI_Recv(&recvbuf, 1, get_mpi_type<T>(), source, tag, comm, MPI_STATUS_IGNORE);
+      }
+    } else { // dest == source
+      recvbuf = sendbuf;
+      this->Barrier(); // Barrier emulates the blocking behavior of send/recv
+    }
+  }
+};
+
+template<typename T>
+void Comm::send_and_recv(std::vector<T>& recvbuf,
+                         int dest,
+                         std::vector<T>& sendbuf,
+                         int source) const
+{
+  if (this->active()) {
+    if (dest != source) {
+      // Send the size of the vector from the source
+      auto n = sendbuf.size();
+      send_and_recv(n, dest, source);
+      // Resize the vector on dest
+      if (rank == dest && recvbuf.size() != n) {
+        recvbuf.resize(n);
+      }
+
+      // Send the vector
+      int tag = source;
+      if (rank == source) {
+        MPI_Send(sendbuf.data(), n, get_mpi_type<T>(), dest, tag, comm);
+      } else if (rank == dest) {
+        MPI_Recv(
+          recvbuf.data(), n, get_mpi_type<T>(), source, tag, comm, MPI_STATUS_IGNORE);
+      }
+    } else { // dest == source
+      recvbuf.resize(sendbuf.size());
+      std::copy(sendbuf.cbegin(), sendbuf.cend(), recvbuf.begin());
+      this->Barrier(); // Barrier emulates the blocking behavior of send/recv
+    }
+  }
+}
+
+template<typename T, size_t N>
+void Comm::send_and_recv(xt::xtensor<T, N>& recvbuf,
+                         int dest,
+                         xt::xtensor<T, N>& sendbuf,
+                         int source) const
+{
+  if (this->active()) {
+    if (dest != source) {
+      // Make sure the shapes match
+      std::vector<size_t> my_sendbuf_shape(sendbuf.shape().begin(), sendbuf.shape().end());
+      std::vector<size_t> my_recvbuf_shape(recvbuf.shape().begin(), recvbuf.shape().end());
```

Comment:
I think I got lucky since `xcontainer::shape()` returns a reference to a data member.  But if you'd like, we can revise this to eliminate some uncertaintity.  

In the xtensor source, I see that [`xcontainer::shape()`](https://github.com/xtensor-stack/xtensor/blob/31acec1e90bbea6d4bc17af0710a123bd5da6689/include/xtensor/xcontainer.hpp#L382-L386) calls [`xstrided_container::shape_impl()`](https://github.com/xtensor-stack/xtensor/blob/31acec1e90bbea6d4bc17af0710a123bd5da6689/include/xtensor/xcontainer.hpp#L834-L838) and returns a reference to the data member `m_shape`.  The data type of `m_shape` is extensively templated, but in the specializations I quickly inspected, it's some kind of xtensor container.  

So in our case, I think `sendbuf.shape().begin()` and `sendbuf.shape().end()` will refer to the beginning and end of the same object.  Likewise for `recvbuf`.

Is there an additional revision you had in mind?

Reply:
Nope that answers the question. Thanks!